src/Rules/Classes/SealedRule.php

```diff
@@ -0,0 +1,69 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Classes;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\DependencyInjection\RegisteredRule;
+use PHPStan\Node\InClassNode;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\VerbosityLevel;
+use function array_values;
+use function sprintf;
+
+/**
+ * @implements Rule<InClassNode>
+ */
+#[RegisteredRule(level: 0)]
+final class SealedRule implements Rule
+{
+
+	public function getNodeType(): string
+	{
+		return InClassNode::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$classReflection = $node->getClassReflection();
+		if ($classReflection->isEnum()) {
+			return [];
+		}
+
+		$className = $classReflection->getName();
+
+		$parents = array_values($classReflection->getImmediateInterfaces());
+		$parentClass = $classReflection->getParentClass();
+		if ($parentClass !== null) {
+			$parents[] = $parentClass;
+		}
+
+		$errors = [];
+		foreach ($parents as $parent) {
+			$sealedTags = $parent->getSealedTags();
+			foreach ($sealedTags as $sealedTag) {
+				$type = $sealedTag->getType();
+				if ($type->isSuperTypeOf(new ObjectType($className))->yes()) {
+					continue;
+				}
+
+				$errors[] = RuleErrorBuilder::message(
+					sprintf(
+						'%s %s is sealed and only permits %s as subtypes, %s given.',
```

Comment:
Instead of this rule, you should implement this extension https://apiref.phpstan.org/2.1.x/PHPStan.Reflection.AllowedSubTypesClassReflectionExtension.html. Which actually make this `@phpstan-sealed` tag useful for analysis, and also already checks what you're checking here, through AllowedSubTypesRule.